internal/config/state_test.go

```diff
@@ -323,3 +332,241 @@ func TestIsValidImage(t *testing.T) {
+						Version:             "1.0.0",
+						LastGeneratedCommit: "abcd123",
+						APIs: []*API{
+							{
+								Path:          "google/cloud/compute/v1",
+								ServiceConfig: "example_service_config.yaml",
+							},
+						},
+						SourcePaths: []string{
+							"src/example/path",
+						},
+						PreserveRegex: []string{
+							"example-preserve-regex",
+						},
+						RemoveRegex: []string{
+							"example-remove-regex",
+						},
+					},
+					{
+						ID:      "google-cloud-storage",
+						Version: "1.2.3",
+						APIs: []*API{
+							{
+								Path:          "google/storage/v1",
+								ServiceConfig: "storage_service_config.yaml",
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:  "empty-librarianState-yaml",
+			state: &LibrarianState{},
+		},
+		{
+			name:  "invalid_file_name",
+			state: &LibrarianState{},
+		},
+		{
+			name:  "invalid content parser",
+			state: &LibrarianState{},
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			tempDir := t.TempDir()
+			contentParser := func(state *LibrarianState) ([]byte, error) {
+				return yaml.Marshal(state)
+			}
+			if test.name == "invalid_file_name" {
+				filePath := filepath.Join(tempDir, "my\x00file.yaml")
+				err := WriteLibrarianState(contentParser, test.state, filePath)
+				if err == nil {
+					t.Errorf("writeLibrarianState() expected an error but got nil")
+				}
+
+				if g, w := err.Error(), "failed to create librarian state file"; !strings.Contains(g, w) {
+					t.Errorf("got %q, wanted it to contain %q", g, w)
+				}
+				return
+			}
+
+			if test.name == "invalid content parser" {
+				filePath := filepath.Join(tempDir, "state.yaml")
+				invalidContentParser := func(state *LibrarianState) ([]byte, error) {
+					return nil, errors.New("simulated parsing error")
+				}
+				err := WriteLibrarianState(invalidContentParser, test.state, filePath)
+				if err == nil {
+					t.Errorf("writeLibrarianState() expected an error but got nil")
+				}
+
+				if g, w := err.Error(), "failed to convert state to bytes"; !strings.Contains(g, w) {
+					t.Errorf("got %q, wanted it to contain %q", g, w)
+				}
+				return
+			}
+
+			filePath := filepath.Join(tempDir, "state.yaml")
+			err := WriteLibrarianState(contentParser, test.state, filePath)
+
+			if err != nil {
+				t.Fatalf("writeLibrarianState() unexpected error: %v", err)
+			}
+
+			// Verify the file content
+			gotBytes, err := os.ReadFile(filePath)
+			if err != nil {
+				t.Fatalf("Failed to read generated file: %v", err)
+			}
+
+			fileName := fmt.Sprintf("%s.yaml", test.name)
+			wantBytes, readErr := os.ReadFile(filepath.Join("..", "..", "testdata", fileName))
+			if readErr != nil {
+				t.Fatalf("Failed to read expected state for comparison: %v", readErr)
+			}
+
+			if diff := cmp.Diff(string(wantBytes), string(gotBytes)); diff != "" {
+				t.Errorf("Generated YAML mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func copyFile(dst, src string) (err error) {
```

Comment:
If the top of this function had the following, then callers wouldn't need to annotate the error:
```
defer func() {
    if err != nil { err = fmt.Errorf("copyFile(%q, %q): %w", dst, src, err) }
}()
```

Reply:
Done